tests/test-umockdev-vala.vala

```diff
@@ -855,6 +855,86 @@ t_mt_uevent ()
   assert_cmpuint (change_count, CompareOperator.EQ, num_changes);
 }
 
+static bool
+ioctl_custom_handle_ioctl_cb(UMockdev.IoctlBase handler, UMockdev.IoctlClient client)
+{
+    if (client.request == 1) {
+      client.complete(*(long*)client.arg.data, 0);
+    } else if (client.request == 2) {
+      client.complete(-1, Posix.ENOMEM);
+    } else {
+      client.complete(-5, 1);
+    }
+    return true;
+}
+
+static bool
+ioctl_custom_handle_write_cb(UMockdev.IoctlBase handler, UMockdev.IoctlClient client)
+{
+    client.set_data("written", client.arg);
+    client.complete(client.arg.data.length, 0);
+    return true;
+}
+
+static bool
+ioctl_custom_handle_read_cb(UMockdev.IoctlBase handler, UMockdev.IoctlClient client)
+{
+    UMockdev.IoctlData written = client.get_data("written");
+
+    assert(written != null);
+    assert(written.data.length == client.arg.data.length);
+
+    Posix.memcpy(client.arg.data, written.data, client.arg.data.length);
```

Comment:
I changed it to return the start of the buffer. It still discards the full thing now, but added a comment to say that this will not usually make sense.